util/tls/tls.go

```diff
@@ -69,25 +79,103 @@ func getTLSVersionByString(version string) (uint16, error) {
-		minVersion, err := getTLSVersionByString(minVersionStr)
-		if err != nil {
-			return nil, err
+// Return array of strings representing TLS versions
+func tlsVersionsToStr(versions []uint16) []string {
+	ret := make([]string, 0)
+	for _, v := range versions {
+		switch v {
+		case tls.VersionTLS10:
+			ret = append(ret, "1.0")
+		case tls.VersionTLS11:
+			ret = append(ret, "1.1")
+		case tls.VersionTLS12:
+			ret = append(ret, "1.2")
+		case tls.VersionTLS13:
+			ret = append(ret, "1.3")
+		default:
+			ret = append(ret, "unknown")
 		}
-		maxVersion, err := getTLSVersionByString(maxVersionStr)
+	}
+	return ret
+}
+
+func getTLSConfigCustomizer(minVersionStr, maxVersionStr, tlsCiphersStr string) (ConfigCustomizer, error) {
+	minVersion, err := getTLSVersionByString(minVersionStr)
+	if err != nil {
+		return nil, err
+	}
+	maxVersion, err := getTLSVersionByString(maxVersionStr)
+	if err != nil {
+		return nil, err
+	}
+	if minVersion > maxVersion {
+		return nil, fmt.Errorf("Minimum TLS version %s must not be higher than maximum TLS version %s", minVersionStr, maxVersionStr)
+	}
+
+	// Cipher suites for TLSv1.3 are not configurable
+	if minVersion == tls.VersionTLS13 {
+		if tlsCiphersStr != DefaultTLSCipherSuite {
+			log.Warnf("TLSv1.3 cipher suites are not configurable, ignoring value of --tlsciphers")
+		}
+		tlsCiphersStr = ""
+	}
+
+	if tlsCiphersStr == "list" {
+		fmt.Printf("Supported TLS ciphers:\n")
+		for _, s := range tls.CipherSuites() {
+			fmt.Printf("* %s (TLS versions: %s)\n", tls.CipherSuiteName(s.ID), strings.Join(tlsVersionsToStr(s.SupportedVersions), ", "))
+		}
+		os.Exit(0)
+	}
+
+	var cipherSuites []uint16
+	if tlsCiphersStr != "" {
+		cipherSuites, err = getTLSCipherSuitesByString(tlsCiphersStr)
 		if err != nil {
 			return nil, err
 		}
-		return func(config *tls.Config) {
-			config.MinVersion = minVersion
-			config.MaxVersion = maxVersion
-		}, nil
+	} else {
+		cipherSuites = make([]uint16, 0)
+	}
+
+	return func(config *tls.Config) {
+		config.MinVersion = minVersion
+		config.MaxVersion = maxVersion
+		config.CipherSuites = cipherSuites
+	}, nil
+
+}
+
+// Adds TLS server related command line options to a command and returns a TLS
+// config customizer object, set up to the options specified
+func AddTLSFlagsToCmd(cmd *cobra.Command) func() (ConfigCustomizer, error) {
+	minVersionStr := ""
+	maxVersionStr := ""
+	tlsCiphersStr := ""
+	cmd.Flags().StringVar(&minVersionStr, "tlsminversion", DefaultTLSMinVersion, "The minimum SSL/TLS version that is acceptable (one of: 1.0|1.1|1.2|1.3)")
+	cmd.Flags().StringVar(&maxVersionStr, "tlsmaxversion", DefaultTLSMaxVersion, "The maximum SSL/TLS version that is acceptable (one of: 1.0|1.1|1.2|1.3)")
+	cmd.Flags().StringVar(&tlsCiphersStr, "tlsciphers", DefaultTLSCipherSuite, "The list of acceptable ciphers to be used when establishing TLS connections. Use 'list' to list available ciphers.")
+
+	return func() (ConfigCustomizer, error) {
```

Comment:
I have refactored the inline func to `getTLSConfigCustomizer()` for better testability.